### ISSUE TITLE
Run the classifiers 

### ISSUE THREAD

**Issue by BartDriesse** [BartDriesse/Thesis#2]
# Import the required modules
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.grid_search import GridSearchCV
from sklearn import cross_validation
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.svm import LinearSVC
from sklearn.linear_model import LogisticRegression

# Read the first csv file 
csv_dataset = 'datasetBAEDA.csv'
raw_df = pd.read_csv(csv_dataset, index_col=0, parse_dates=[2], infer_datetime_format=True)
raw_df.head()

# Now group the data by year in Start.Data and plot it
yearly_sum = raw_df.groupby(raw_df['Start.Date'].map(lambda x: x.year)).sum()
# Keep only data between 2004 to 2016
selected_data = yearly_sum.ix[(yearly_sum.index > 2003) & (yearly_sum.index < 2017), 'Converted']
# Plot Converted vs. the year
plt.plot(selected_data)
plt.grid(), plt.xlabel('Starting Year'), plt.ylabel('Converted'), plt.title('Number of Students Converted each Year')
plt.show()

# Load the second CSV file
second_dataset = 'bachelor.csv'
raw_df_2 = pd.read_csv(second_dataset, index_col=0)
# Combine the 2 IDs columns - Contact.ID and Lead.ID.Marketing.Cloud
raw_df_2['Contact.ID'] = raw_df_2['Contact.ID'].combine_first(raw_df_2['Lead.ID.Marketing.Cloud'])
# Drop the Lead.ID.Marketing.Cloud column as it is NO more needed
raw_df_2.drop('Lead.ID.Marketing.Cloud', inplace=True, axis=1)

final_df = raw_df_2

# Draw the sum of the features
brochure = sum(final_df['Profile.Brochure.Count'] > 0)
meeloopdagen = sum(final_df['Profile.Meeloopdagen.Ba.Count'] > 0)
onderwijsbeurs = sum(final_df['Profile.Onderwijsbeurs.Count'] > 0)
open_dag_ba = sum(final_df['Profile.Open.Dag.Ba.Count'] > 0)
case = sum(final_df['Profile.Case.Count'] > 0)
webinar = sum(final_df['Profile.Webinar.Count'] > 0)

# Some parameters for chart
counts_list = [brochure, meeloopdagen,
               onderwijsbeurs, open_dag_ba, case, webinar]
feat_labels = ('Brochure', 'Meeloopdagen', 'Onderwijsbeurs', 'Open Dag', 'Case', 'Webinar')
ind = np.arange(len(counts_list))  # the x locations for the groups
width = 0.5                        # the width of the bars
# Make the chart now
fig, ax = plt.subplots()
ax.bar(ind, counts_list, width=0.5, align='center')
ax.set_ylabel('Number of students')
ax.set_title('How many students visit the marketing events?')
ax.set_xticks(ind)
ax.set_xticklabels(feat_labels, rotation=90)
plt.show()

# Prepare X and Y data
x_cols_list = [
    'Profile.Brochure.Count',
    'Profile.Meeloopdagen.Ba.Count', 'Profile.Onderwijsbeurs.Count', 'Profile.Open.Dag.Ba.Count', 
    'Profile.Case.Count', 'Profile.Webinar.Count'
]
x_data = final_df.ix[:, x_cols_list]
y_data = final_df.ix[:, 'Converted'].as_matrix().ravel()
#split into train and test data
X_train, X_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.20, random_state=30)

# Run the lineair svm on default settings
lsvc1= LinearSVC()
lsvc1.fit(X_train, y_train)
all_scores = cross_validation.cross_val_score(lsvc1, X_train, y_train, cv=5)
#Print all the scores of the cross_validation
#print(all_scores)
print("Mean score of the Linear SVM Classifier: ", np.mean(all_scores)*100, "%")

# Run the logistic regression on default settings
lsvc2= LogisticRegression()
lsvc2.fit(X_train, y_train)
all_scores = cross_validation.cross_val_score(lsvc2,X_train, y_train, cv=5)
#Print all the scores of the cross_validation
#print(all_scores)
print("Mean score of the logstics regression: ", np.mean(all_scores)*100, "%")

   
# Grid search for linear svm classifier
clfLRl2 = LinearSVC()
param_grid = {'C': [0.0001, 0.001, 0.01, 0.1, 1.0, 10.0, 100.0]}
clf = GridSearchCV(clfLRl2, param_grid, cv=5)
clf = clf.fit(X_train, y_train)
print("SVM", clf.best_score_)
print( clf.best_params_)
# Grid search for Logistic regression
clf = LogisticRegression()
param_grid = {'C': [0.0001, 0.001, 0.01, 0.1, 1.0,10.0, 100.0] }
clf = GridSearchCV(clf, param_grid, cv=5)
clf = clf.fit(X_train, y_train)
print( "Logistic regression: ",clf.best_score_)
print( clf.best_params_)
# Parameters to search for in Grid Search

parameters = {
    'C' : (0.0001, 0.001, 0.01, 0.1, 1.0, 100.0)
}

# Train the classifiers with the optimal C and report accuracy

# Do Grid Search to find the best parameters
searched_clfLOG = GridSearchCV(LogisticRegression(), param_grid=parameters, cv=5, n_jobs=1)
best_clfLOG = searched_clfLOG.fit(X=X_train, y=y_train)

# Get the best score & best parameters
print("Best Score: %.3f%%" % (best_clfLOG.best_score_*100))
print("Best Parameters:")
for param_name in sorted(parameters.keys()):
    print("\t%s: %r" % (param_name, best_clfLOG.best_params_[param_name]))

# Now use the best parameters to train and test the classifier
final_clfLOG = LogisticRegression(
    C=best_clfLOG.best_params_['C']
)

# Firstly use all features to compute the score
all_scores = cross_validation.cross_val_score(final_clfLOG, X=X_train, y=y_train, cv=5, n_jobs=1)
print("Results with ALL feats!!")
print(all_scores)
print("Mean Score of logistic Regression: " + str(all_scores.mean()*100) + "%")

# Now drop each feature one by one and see effect of that feature on accuracy
for i in range(0,len(x_cols_list)):
    excluded_feat = x_cols_list[i]
    excluded_list = [x for j,x in enumerate(x_cols_list) if j != i]
    new_x_data = final_df.ix[:, excluded_list]
    print("---------------------------------------------------")
    print("Results of logistic Regression with " + str(excluded_feat) + " excluded!!")
    print("---------------------------------------------------")
    all_scores = cross_validation.cross_val_score(final_clfLOG, X=new_x_data, y=y_data, cv=5, n_jobs=1)
    print(all_scores)
    print("Mean Score of logistic Regression: " + str(all_scores.mean()*100) + "%")
    print("")

# Run gridsearch on svm

searched_clfSVM = GridSearchCV(LinearSVC(), param_grid=parameters, cv=5, n_jobs=1)
best_clfSVM = searched_clfSVM.fit(X=X_train, y=y_train)

# Get the best score & best parameters
print("Best Score: %.3f%%" % (best_clfSVM.best_score_*100))
print("Best Parameters:")
for param_name in sorted(parameters.keys()):
    print("\t%s: %r" % (param_name, best_clfSVM.best_params_[param_name]))

# Now use the best parameters to train and test the classifier
final_clfSVM = LinearSVC(
    C=best_clfSVM.best_params_['C']
)

# Firstly use all features to compute the score
all_scores = cross_validation.cross_val_score(final_clfSVM, X=X_train, y=y_train, cv=5, n_jobs=1)
print("Results with ALL feats!!")
print(all_scores)
print("Mean Score of SVM: " + str(all_scores.mean()*100) + "%")

# Now drop each feature one by one and see effect of that feature on accuracy
for i in range(0,len(x_cols_list)):
    excluded_feat = x_cols_list[i]
    excluded_list = [x for j,x in enumerate(x_cols_list) if j != i]
    new_x_data = final_df.ix[:, excluded_list]
    print("---------------------------------------------------")
    print("Results for SVM with " + str(excluded_feat) + " excluded!!")
    print("---------------------------------------------------")
    all_scores = cross_validation.cross_val_score(final_clfSVM, X=new_x_data, y=y_data, cv=5, n_jobs=1)
    print(all_scores)
    print("Mean Score of SVM: " + str(all_scores.mean()*100) + "%")
    print("")

model = final_clfSVM
model.fit(X_train, y_train)
y_test_pred = model.predict(X_test)

print('SVM', accuracy_score(y_test, y_test_pred)) 
print(classification_report(y_test, y_test_pred)) # precision and recall
print(confusion_matrix(y_test, y_test_pred))

model1 = final_clfLOG
model1.fit(X_train, y_train)
y_test_pred = model1.predict(X_test)

print('Logistic regression', accuracy_score(y_test, y_test_pred)) 
print(classification_report(y_test, y_test_pred)) # precision and recall
print(confusion_matrix(y_test, y_test_pred))

# Show the coefficients
coefficients = pd.concat([pd.DataFrame(X_train.columns),pd.DataFrame(np.transpose(final_clfSVM.coef_))], axis = 1)
print(coefficients)
#print(lsvc1.coef_)
print("Number of iterations: ",final_clfSVM.n_iter_)
print("Intercept of the model: ",final_clfSVM.intercept_)

coefficients2 = pd.concat([pd.DataFrame(X_train.columns),pd.DataFrame(np.transpose(final_clfLOG.coef_))], axis = 1)
print(coefficients2)
#print(lsvc1.coef_)
print("Number of iterations: ",final_clfLOG.n_iter_)
print("Intercept of the model: ",final_clfLOG.intercept_)**
